<!DOCTYPE html>
<html class="writer-html5" lang="English">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Various utility scripts / functions &mdash; FitHiChIP 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Frequently asked question" href="FAQ.html" />
    <link rel="prev" title="Differential analysis of HiChIP loops" href="DiffLoops.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            FitHiChIP
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Setting up configuration file</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Details of FitHiChIP outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="DiffLoops.html">Differential analysis of HiChIP loops</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Various utility scripts / functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#generating-contact-matrices-of-different-resolution-bin-size">1. generating contact matrices of different resolution (bin size)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#inferring-peaks-from-hichip-data-for-use-in-the-hichip-pipeline">2. inferring peaks from HiChIP data (for use in the HiChIP pipeline)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#merging-a-set-of-chip-seq-alignment-files">3. merging a set of ChIP-seq alignment files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#producing-fithichip-loops-with-different-fdr-thresholds">4. Producing FitHiChIP loops with different FDR thresholds</a></li>
<li class="toctree-l2"><a class="reference internal" href="#applying-merge-filtering-on-any-set-of-significant-loops-generated-by-fithichip-or-any-other-loop-calling-method">5. Applying merge filtering on any set of significant loops (generated by FitHiChIP or any other loop calling method)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualization-of-the-significant-loops-in-washu-epigenome-browser">6. Visualization of the significant loops in WashU epigenome browser</a></li>
<li class="toctree-l2"><a class="reference internal" href="#simulating-hichip-contact-matrix-from-hic-and-chip-seq-data">7. Simulating HiChIP contact matrix from HiC and ChIP-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#applying-fithichip-on-capture-hic-data">8. Applying FitHiChIP on capture HiC data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently asked question</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">FitHiChIP</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Various utility scripts / functions</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/usage/Utilities.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="various-utility-scripts-functions">
<h1>Various utility scripts / functions<a class="headerlink" href="#various-utility-scripts-functions" title="Permalink to this heading"></a></h1>
<section id="generating-contact-matrices-of-different-resolution-bin-size">
<h2>1. generating contact matrices of different resolution (bin size)<a class="headerlink" href="#generating-contact-matrices-of-different-resolution-bin-size" title="Permalink to this heading"></a></h2>
<p>The script <strong>ContactMatDiffRes.sh</strong> within the folder <strong>Imp_Scripts</strong> generates contact matrix of desired bin size given a valid pairs file generated from the HiC-pro pipeline. Parameters associated with this script is:</p>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-V <var>ValidPairsFile</var></span></kbd></dt>
<dd><p>Name of the valid pairs file generated from the HiC-pro pipeline</p>
</dd>
<dt><kbd><span class="option">-B <var>BinSize</var></span></kbd></dt>
<dd><p>Size of the bin (in bytes: target resolution): default 5000 (5 Kb)</p>
</dd>
<dt><kbd><span class="option">-C <var>ChrSizeFile</var></span></kbd></dt>
<dd><p>File containing the size of the chromosomes corresponding to the reference genome
(such as hg19.chrom.sizes)</p>
</dd>
<dt><kbd><span class="option">-D <var>OutDir</var></span></kbd></dt>
<dd><p>Output directory which will contain the contact matrix of the target resolution</p>
</dd>
<dt><kbd><span class="option">-H <var>HiCProDir</var></span></kbd></dt>
<dd><p>Directory containing the HiC-pro installed package.</p>
</dd>
</dl>
</div></blockquote>
<p>A folder <strong>Matrix_BinSize${BinSize}</strong> will be created under the specified output directory. The bin specific
interval file name is <strong>Matrix_abs.bed</strong> and the corresponding contact matrix file name is <strong>Matrix.matrix</strong></p>
</section>
<section id="inferring-peaks-from-hichip-data-for-use-in-the-hichip-pipeline">
<h2>2. inferring peaks from HiChIP data (for use in the HiChIP pipeline)<a class="headerlink" href="#inferring-peaks-from-hichip-data-for-use-in-the-hichip-pipeline" title="Permalink to this heading"></a></h2>
<p>The script <strong>PeakInferHiChIP.sh</strong> within the folder <strong>Imp_Scripts</strong> is used to infer peaks from HiChIP data. The script can be used if HiC-pro pipeline is already executed on a given pair of reads (such as .fastq.gz read pairs). The script uses <strong>macs2</strong> package for inferring the peaks. Parameters associated with this script are as follows:</p>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-H <var>HiCProDir</var></span></kbd></dt>
<dd><p>Directory containing the reads generated by HiC-pro pipeline.
Within this directory, files of the formats .ValidPairs, .DEPairs, .REPairs, and .SCPairs
are present, which corresponds to different categories of reads generated by the HiC-pro pipeline.</p>
</dd>
<dt><kbd><span class="option">-D <var>OutDir</var></span></kbd></dt>
<dd><p>Directory to contain the output set of peaks. Default: current directory</p>
</dd>
<dt><kbd><span class="option">-R <var>refGenomeStr</var></span></kbd></dt>
<dd><p>Reference genome specific string used for MACS2. Default is 'hs' for human chromosome.
For mouse, specify 'mm'.</p>
</dd>
<dt><kbd><span class="option">-M <var>MACS2ParamStr</var></span></kbd></dt>
<dd><p>String depicting the parameters for MACS2. Default: &quot;--nomodel --extsize 147 -q 0.01&quot;</p>
</dd>
<dt><kbd><span class="option">-L <var>ReadLength</var></span></kbd></dt>
<dd><p>Length of reads for the HiC-pro generated reads. Default 75</p>
</dd>
</dl>
</div></blockquote>
<p>The script uses all of the DE, SC, RE and validpairs reads generated from the HiC-pro pipeline to infer peaks.
The folder <strong>MACS2_ExtSize</strong> within the specified output directory contains the MACS2 generated peaks.</p>
</section>
<section id="merging-a-set-of-chip-seq-alignment-files">
<h2>3. merging a set of ChIP-seq alignment files<a class="headerlink" href="#merging-a-set-of-chip-seq-alignment-files" title="Permalink to this heading"></a></h2>
<p>The script <strong>MergeBAMInferPeak.sh</strong> within the folder <strong>Imp_Scripts</strong> processes a set of ChIP-seq alignment files, to generate a merged alignment, infer peak (using MACS2) from the generated alignment, and also determines the coverage of individual bins (according to a specified fixed bin size) with respect to individual input ChIP-seq alignment files. Output of this script is used as an input to the differential analysis module (mentioned next). Parameters associated with this script are as follows:</p>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-I <var>InpFile</var></span></kbd></dt>
<dd><p>A list of ChIP-seq BAM files. Multiple BAM files are to be mentioned in the
format: -I bamfile1.bam -I bamfile2.bam -I bamfile3.bam and so on</p>
</dd>
<dt><kbd><span class="option">-D <var>OutBaseDir</var></span></kbd></dt>
<dd><p>Directory containing the output set of peaks. Default: current directory</p>
</dd>
<dt><kbd><span class="option">-R <var>refGenome</var></span></kbd></dt>
<dd><p>Reference genome string used for MACS2. Default is 'hs' for human chromosome. For mouse, specify 'mm'</p>
</dd>
<dt><kbd><span class="option">-M <var>MACS2ParamStr</var></span></kbd></dt>
<dd><p>String depicting the parameters for MACS2. Default: &quot;--nomodel --extsize 147 -q 0.01&quot;</p>
</dd>
<dt><kbd><span class="option">-C <var>ChrSizeFile</var></span></kbd></dt>
<dd><p>Filename containing the chromosome size information for the reference genome</p>
</dd>
<dt><kbd><span class="option">-b <var>BinSize</var></span></kbd></dt>
<dd><p>BinSize in base pair. Used to compute the ChIP-seq coverage for individual input
BAM files. Default = 5000 (means 5 Kb)</p>
</dd>
</dl>
</div></blockquote>
<p>Output files generated from this script:</p>
<blockquote>
<div><ol class="arabic">
<li><p><strong>${OutBaseDir}/merged_input.bam</strong>: merged ChIP-seq alignment file</p></li>
<li><p><strong>${OutBaseDir}/MACS2_Out/out.macs2_peaks.narrowPeak</strong>: Peak (derived by MACS2) file generated by merged ChIP-seq alignment.</p></li>
<li><p><strong>${OutBaseDir}/ChIPCoverage1.bed</strong>, <strong>${OutBaseDir}/ChIPCoverage2.bed</strong>, ..., <strong>${OutBaseDir}/ChIPCoverageN.bed</strong>,</p>
<blockquote>
<div><p>where <strong>N</strong> is the number of input ChIP-seq alignment files. Each of these output files contain the coverage of corresponding input alignment file (with respect to the specified bin size).</p>
</div></blockquote>
</li>
</ol>
</div></blockquote>
</section>
<section id="producing-fithichip-loops-with-different-fdr-thresholds">
<h2>4. Producing FitHiChIP loops with different FDR thresholds<a class="headerlink" href="#producing-fithichip-loops-with-different-fdr-thresholds" title="Permalink to this heading"></a></h2>
<p>It is useful to obtain significant loops on different FDR threholds, without running the complete FitHiChIP module with different FDR threshold parameters.</p>
<p>For a custom FDR threshold of 0.05, first user should apply the following awk script on the file PREFIX.interactions_FitHiC.bed (mentioned in section 4.1.4, file no C):</p>
<blockquote>
<div><p><strong>awk '((NR==1) || ($NF&lt;0.05))' ${PREFIX}.interactions_FitHiC.bed &gt; ${PREFIX}.interactions_FitHiC_Q0.05.bed</strong></p>
</div></blockquote>
</section>
<section id="applying-merge-filtering-on-any-set-of-significant-loops-generated-by-fithichip-or-any-other-loop-calling-method">
<h2>5. Applying merge filtering on any set of significant loops (generated by FitHiChIP or any other loop calling method)<a class="headerlink" href="#applying-merge-filtering-on-any-set-of-significant-loops-generated-by-fithichip-or-any-other-loop-calling-method" title="Permalink to this heading"></a></h2>
<p>It is useful to apply the merge filtering on any set of significant loops, not just of FitHiChIP.</p>
<p>Assuming the source code of FitHiChIP is placed within the folder <strong>${CodeDir}</strong>, user can merge the adjacent loops using the following command:</p>
<blockquote>
<div><p>python ${CodeDir}/src/CombineNearbyInteraction.py \</p>
<p class="attribution">—InpFile ${PREFIX}.interactions_FitHiC_Q0.05.bed \</p>
</div></blockquote>
<blockquote>
<div><p>--OutFile ${PREFIX}.interactions_FitHiC_Q0.05_MergeNearContacts.bed \</p>
<p class="attribution">—headerInp 1 --binsize 5000 --percent 100 --Neigh 2</p>
</div></blockquote>
<blockquote>
<div><p>where, the options are:</p>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">--InpFile <var>InpFile</var></span></kbd></dt>
<dd><p>Input file containing FitHiChIP significant interactions.</p>
</dd>
<dt><kbd><span class="option">--OutFile <var>OutFile</var></span></kbd></dt>
<dd><p>Output file to contain the significant and merge filtered loops.</p>
</dd>
<dt><kbd><span class="option">--headerInp <var>Header</var></span></kbd></dt>
<dd><p>Boolean value (0/1): 1 means InpFile has header line. 0 means no header line is in the
input loop file.</p>
</dd>
<dt><kbd><span class="option">--binsize <var>BinSize</var></span></kbd></dt>
<dd><p>Bin size (in bp). For example, 5000 means 5 Kb bin size.</p>
</dd>
<dt><kbd><span class="option">--percent <var>PctVal</var></span></kbd></dt>
<dd><p>% of significant loops considered for merge filtering (default PctVal = 100).
Users are advised to keep this value.</p>
</dd>
<dt><kbd><span class="option">--Neigh <var>Val</var></span></kbd></dt>
<dd><p>Window size (number of bins) for merge filtering. Default is Val = 2, means
2X2 bins are employed as window. Users are advised to keep this value.</p>
</dd>
</dl>
</div></blockquote>
</div></blockquote>
<p>The file <strong>${PREFIX}.interactions_FitHiC_Q0.05_MergeNearContacts.bed</strong> is the set of loops generated after merging adjacent ones.</p>
</section>
<section id="visualization-of-the-significant-loops-in-washu-epigenome-browser">
<h2>6. Visualization of the significant loops in WashU epigenome browser<a class="headerlink" href="#visualization-of-the-significant-loops-in-washu-epigenome-browser" title="Permalink to this heading"></a></h2>
<p>Check the page <a class="reference internal" href="output.html#details-of-fithichip-outputs"><span class="std std-ref">Details of FitHiChIP outputs</span></a> for a detailed description of FitHiChIP output signficant loops and their visualization in WashU epigenome browser.</p>
</section>
<section id="simulating-hichip-contact-matrix-from-hic-and-chip-seq-data">
<h2>7. Simulating HiChIP contact matrix from HiC and ChIP-seq data<a class="headerlink" href="#simulating-hichip-contact-matrix-from-hic-and-chip-seq-data" title="Permalink to this heading"></a></h2>
<p>In FitHiChIP manuscript, we have checked its robustness by simulating HiChIP data from reference HiC and ChIP-seq data. Here we provide a brief description of such simulation procedure.</p>
<p>The file Imp_Scripts/HiC_Simulate_by_ChIP_Coverage.r contains the simulation procedure.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>The simulation procedure and corresponding implementation is still very basic.</p></li>
<li><p>The generated HiChIP contact map is shown to exhibit high bias for low range contacts. So, user may use this implementation as a basic template, and may incorporate their own modifications.</p></li>
</ul>
</div>
<p>The script is to be executed as</p>
<p>Rscript Imp_Scripts/HiC_Simulate_by_ChIP_Coverage.r <strong>[OPTIONS]</strong></p>
<p>Where, <strong>[OPTIONS]</strong> are the following:</p>
<dl class="option-list">
<dt><kbd><span class="option">--chr</span></kbd></dt>
<dd><ul class="simple">
<li><p>Comma or colon separated list of chromosome names for which the simulation would be carried out. Its value can be a single chromosome like chr1, or a list of chromosomes like chr1,chr2</p></li>
<li><p>If nothing is provided, all chromosomes within the specified chromosome size file of the reference genome would be analyzed.</p></li>
</ul>
</dd>
<dt><kbd><span class="option">--OutDir</span></kbd></dt>
<dd><ul class="simple">
<li><p>Output directory to contain the simulated results. <strong>Mandatory parameter</strong>.</p></li>
</ul>
</dd>
<dt><kbd><span class="option">--ChrSizeFile</span></kbd></dt>
<dd><ul class="simple">
<li><p>File having the chromosome sizes for the reference genome.</p></li>
<li><p><strong>Mandatory parameter</strong>.</p></li>
</ul>
</dd>
<dt><kbd><span class="option">--ChIPAlignFile</span></kbd></dt>
<dd><ul class="simple">
<li><p>ChIP-seq alignment file. <strong>Mandatory parameter</strong>.</p></li>
<li><p>Can either be in BAM or Bedgraph (4-column) format**.</p></li>
</ul>
</dd>
<dt><kbd><span class="option">--HiCMapFile</span></kbd></dt>
<dd><ul class="simple">
<li><p>Reference HiC contact map file, having the following fields in tab delimited format: chr1, start1, end1, chr2, start2, end2, cc.</p></li>
<li><p>Here the triplet (chr,start,end) corresponds to one interacting interval. The field cc denotes the contact count between corresponding interacting regions. It can be either raw contact count, or normalized (by ICE).</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>Output HiChIP would have the same resolution (bin size) as the input HiC map.</p></li>
</ul>
</div>
</dd>
<dt><kbd><span class="option">--TotalRead</span></kbd></dt>
<dd><ul class="simple">
<li><p>Total coverage of the output simulated HiChIP contact matrix. That is, total contact count in the simulated HiChIP contact map would be equal to this integer value.</p></li>
<li><p>Default is 1000000, means the output simulated map would have 1M reads.</p></li>
</ul>
</dd>
</dl>
<p>Within the specified output folder <strong>OutDir</strong>, folders of the following name are created:</p>
<blockquote>
<div><p>Simulate_HiChIP_chr_*chrName*_B_*BinSize*</p>
<p>where, <em>chrName</em> corresponds to one chromosome within the list of chromosomes specified, and <em>BinSize</em> corresponds to the bin size (resolution) of the reference HiC loops.</p>
<p>Within this folder, the file <strong>Dumped_Balanced_CC_RegrFactor.bed</strong> contains the simulated HiChIP contact map for the corresponding chromosome.</p>
<blockquote>
<div><ul class="simple">
<li><p>Its first 6 fields are the interacting regions</p></li>
<li><p>7th field is the raw contact count obtained by the ChIP-seq coverage values</p></li>
<li><p>8th (last) field is the scaled contact count according to the parameter --TotalRead. This contact count is to be used for application, or for visualization.</p></li>
</ul>
</div></blockquote>
</div></blockquote>
</section>
<section id="applying-fithichip-on-capture-hic-data">
<h2>8. Applying FitHiChIP on capture HiC data<a class="headerlink" href="#applying-fithichip-on-capture-hic-data" title="Permalink to this heading"></a></h2>
<p>FitHiChIP can be applied on the Promoter Capture HiC assay as well.</p>
<p>1) First, user needs to run HiC-pro using the command:
./HiC-Pro -i data -o output -c config -s mapping -s proc_hic -s quality_checks -s merge_persample</p>
<p>(as suggested in <a class="reference external" href="https://github.com/nservant/HiC-Pro/issues/102">https://github.com/nservant/HiC-Pro/issues/102</a>)</p>
<p>This will generate allValidPairs file with the list of valid interactions.</p>
<ol class="arabic simple" start="2">
<li><p>Then, use the design bait file of the promoter capture HiC data, and extract the corresponding intervals (chromosome, start, end) in a separate file X. This file X will be used as the peak file in FitHiChIP execution (via the configuration parameter &quot;PeakFile=&quot;)</p></li>
</ol>
<p>We have used peak to all interactions (IntType=3) using coverage bias regression when executing FitHiChIP on promoter capture HiC data.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="DiffLoops.html" class="btn btn-neutral float-left" title="Differential analysis of HiChIP loops" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="FAQ.html" class="btn btn-neutral float-right" title="Frequently asked question" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, Sourya Bhattacharyya, Ferhat Ay.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>